query-projector/src/pull.rs

```diff
@@ -0,0 +1,126 @@
+    Puller,
+};
+
+use errors::{
+    Result,
+};
+
+use super::{
+    Index,
+    rusqlite,
+};
+
+#[derive(Clone, Debug)]
+pub(crate) struct PullOperation(pub(crate) Vec<PullAttributeSpec>);
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct PullIndices {
+    pub(crate) sql_index: Index,                   // SQLite row index.
+    pub(crate) output_index: usize,
+}
+
+impl PullIndices {
+    fn zero() -> PullIndices {
+        PullIndices {
+            sql_index: 0,
+            output_index: 0,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct PullTemplate {
+    pub(crate) indices: PullIndices,
+    pub(crate) op: PullOperation,
+}
+
+pub(crate) struct PullConsumer<'schema> {
+    indices: PullIndices,
+    schema: &'schema Schema,
+    puller: Puller,
+    entities: BTreeSet<Entid>,
+    results: BTreeMap<Entid, ValueRc<StructuredMap>>,
+}
+
+impl<'schema> PullConsumer<'schema> {
+    pub(crate) fn for_puller(puller: Puller, schema: &'schema Schema, indices: PullIndices) -> PullConsumer<'schema> {
+        PullConsumer {
+            indices: indices,
+            schema: schema,
+            puller: puller,
+            entities: Default::default(),
+            results: Default::default(),
+        }
+    }
+
+    pub(crate) fn for_template(schema: &'schema Schema, template: &PullTemplate) -> Result<PullConsumer<'schema>> {
+        let puller = Puller::prepare(schema, template.op.0.clone())?;
+        Ok(PullConsumer::for_puller(puller, schema, template.indices))
+    }
+
+    pub(crate) fn for_operation(schema: &'schema Schema, operation: &PullOperation) -> Result<PullConsumer<'schema>> {
+        let puller = Puller::prepare(schema, operation.0.clone())?;
+        Ok(PullConsumer::for_puller(puller, schema, PullIndices::zero()))
+    }
+
+    pub(crate) fn collect_entity<'a, 'stmt>(&mut self, row: &rusqlite::Row<'a, 'stmt>) -> Entid {
+        let entity = row.get(self.indices.sql_index);
+        self.entities.insert(entity);
+        entity
+    }
+
+    pub(crate) fn pull(&mut self, sqlite: &rusqlite::Connection) -> Result<()> {
+        let entities: Vec<Entid> = self.entities.iter().cloned().collect();
+        self.results = self.puller.pull(self.schema, sqlite, entities)?;
+        Ok(())
+    }
+
+    pub(crate) fn expand(&self, bindings: &mut [Binding]) {
+        if let Binding::Scalar(TypedValue::Ref(id)) = bindings[self.indices.output_index] {
+            if let Some(pulled) = self.results.get(&id).cloned() {
+                bindings[self.indices.output_index] = Binding::Map(pulled);
+            } else {
+                bindings[self.indices.output_index] = Binding::Map(ValueRc::new(Default::default()));
+            }
+        }
+    }
+
+    // TODO: do we need to include empty maps for entities that didn't match any pull?
+    pub(crate) fn to_coll_results(self) -> Vec<Binding> {
+        self.results.values().cloned().map(|vrc| Binding::Map(vrc)).collect()
+    }
+
+    pub(crate) fn done(&mut self) {
```

Comment:
It's not needed for `to_coll_results`, because that consumes `self`, and so both of these collections are dropped.

I only use `done` here in the tuple projector, where after each column we can discard the pull consumer. I've changed that to just use `into_iter`, which means we can kill `done`.